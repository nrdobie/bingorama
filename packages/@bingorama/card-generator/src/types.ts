export type Card = [number, ...number[]] & { length: 25 }

export type Seed = () => number
