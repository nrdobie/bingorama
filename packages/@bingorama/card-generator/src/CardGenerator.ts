import * as createSeed from 'seed-random'
import { Card, Seed } from './types'
import { compareArrays } from './utilities/compareArrays'
import { getRandomInt } from './utilities/getRandomInt'

export interface ICardGeneratorOptions {
  /**
   * Random seed to use for generating cards, same seed will deliver same cards.
   * When null a random seed will be used.
   */
  seed: string | null

  /**
   * When false generate will check if the card could trigger a double bingo when using classic bingo rules
   */
  allowDuplicates: boolean

  /**
   * Adds a free spaces to the middle of the I column
   */
  includeFreeSpace: boolean

  /**
   * Increases randomness by adding more shuffles to the columns
   */
  shuffleIterations: number

  /**
   * Maximum number of duplicate cards in a row before quitting
   */
  duplicateThreshold: number
}

/**
 * Generates bingo card
 */
export class CardGenerator {
  /**
   * Default options for 
   */
  public static DEFAULT_OPTIONS: ICardGeneratorOptions = {
    seed: null,
    allowDuplicates: false,
    includeFreeSpace: true,
    shuffleIterations: 250,
    duplicateThreshold: 100
  }

  public options: ICardGeneratorOptions

  private numberList = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]

  constructor(options: Partial<ICardGeneratorOptions> = {}) {
    this.options = {
      ...CardGenerator.DEFAULT_OPTIONS,
      ...options
    }
  }

  public async generateCards(count: number): Promise<Card[]> {
    let runningDuplicates = 0
    const cards: Card[] = []
    const seed = createSeed(this.options.seed || Math.random().toString())

    while (cards.length < count) {
      const card = await this.generateCard(seed)

      if (
        this.options.allowDuplicates ||
        !cards.some(a => this.classicMatch(a, card))
      ) {
        runningDuplicates = 0
        cards.push(card)
      } else {
        runningDuplicates++

        if (runningDuplicates > this.options.duplicateThreshold) {
          break
        }
      }
    }

    return cards
  }

  public generateCard(seed: Seed): Promise<Card> {
    return new Promise(resolve => {
      const card = []

      for (let i = 0; i < 5; i++) {
        const numbers = this.getShuffledNumberList(seed)

        // I free space
        if (i === 2 && this.options.includeFreeSpace) {
          card.push(...numbers.slice(0, 2), 0, ...numbers.slice(2, 4))
        } else {
          card.push(...numbers.slice(0, 5))
        }
      }

      setTimeout(() => resolve(card as Card), 0)
    })
  }

  /**
   * Checks if two cards can trigger double bingo with classic bingo rules
   * @param a Card A
   * @param b Card B
   */
  public classicMatch(a: Card, b: Card): boolean {
    // Check for direct match
    if (compareArrays(a, b)) {
      return true
    }

    // Check for column match
    for (let i = 0; i < 5; i++) {
      const start = i * 5
      const end = start + 5

      if (compareArrays(a.slice(start, end), b.slice(start, end))) {
        return true
      }
    }

    // Check for row match
    for (let i = 0; i < 5; i++) {
      const aList = [a[i], a[i + 5], a[i + 10], a[i + 15], a[i + 20]]
      const bList = [b[i], b[i + 5], b[i + 10], b[i + 15], b[i + 20]]

      if (compareArrays(aList, bList)) {
        return true
      }
    }

    // Check for diagonals
    const aListDiagonalRight = [a[0], a[6], a[12], a[18], a[24]]
    const bListDiagonalRight = [b[0], b[6], b[12], b[18], b[24]]

    if (compareArrays(aListDiagonalRight, bListDiagonalRight)) {
      return true
    }

    const aListDiagonalLeft = [a[4], a[8], a[12], a[16], a[20]]
    const bListDiagonalLeft = [b[4], b[8], b[12], b[16], b[20]]

    if (compareArrays(aListDiagonalLeft, bListDiagonalLeft)) {
      return true
    }

    return false
  }

  private getShuffledNumberList(seed: Seed) {
    const shuffledNumberList = [...this.numberList]

    for (let i = 0; i < this.options.shuffleIterations; i++) {
      shuffledNumberList.sort(() => getRandomInt(seed, -1, 2))
    }

    return shuffledNumberList
  }
}
