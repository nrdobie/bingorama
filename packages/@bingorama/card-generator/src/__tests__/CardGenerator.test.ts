import * as createSeed from 'seed-random'
import { CardGenerator } from '../CardGenerator'
import { Card, Seed } from '../types'

describe('CardGenerator', () => {
  let cardGenerator: CardGenerator

  beforeEach(() => {
    cardGenerator = new CardGenerator()
  })

  describe('#generateCards', () => {
    let generateCardSpy: jest.SpyInstance<Promise<Card>, [Seed]>
    let cardA: Card
    let cardB: Card

    beforeEach(() => {
      cardGenerator.options.seed = 'test'

      generateCardSpy = jest.spyOn(cardGenerator, 'generateCard')

      // prettier-ignore
      cardA = [
        1, 2, 3, 4, 5,
        1, 2, 3, 4, 5,
        1, 2, 3, 4, 5,
        1, 2, 3, 4, 5,
        1, 2, 3, 4, 5
      ]

      // prettier-ignore
      cardB = [
        11, 12, 13, 14, 15,
        11, 12, 13, 14, 15,
        11, 12, 13, 14, 15,
        11, 12, 13, 14, 15,
        11, 12, 13, 14, 15
      ]
    })

    it('should generate a card set', async () => {
      await expect(cardGenerator.generateCards(2)).resolves.toMatchSnapshot()
    })

    it('should generate random seeds', async () => {
      cardGenerator.options.seed = null

      const setA = await cardGenerator.generateCards(2)
      const setB = await cardGenerator.generateCards(2)

      expect(setA).not.toEqual(setB)
    })

    it('should remove duplicates', async () => {
      generateCardSpy
        .mockReturnValueOnce(Promise.resolve(cardA))
        .mockReturnValueOnce(Promise.resolve(cardA))
        .mockReturnValueOnce(Promise.resolve(cardB))
      await expect(cardGenerator.generateCards(2)).resolves.toMatchSnapshot()
      expect(generateCardSpy).toHaveBeenCalledTimes(3)
    })

    it('should fail if too many duplicates', async () => {
      generateCardSpy.mockReturnValue(Promise.resolve(cardA))
      await expect(cardGenerator.generateCards(2)).resolves.toMatchSnapshot()
    })
  })

  describe('#generateCard', () => {
    let seed: Seed

    beforeEach(() => {
      seed = createSeed('test')
    })

    it('should create a card that matches snapshot', async () => {
      await expect(cardGenerator.generateCard(seed)).resolves.toMatchSnapshot()
    })

    it('should create a card that matches snapshot with no free space', async () => {
      cardGenerator.options.includeFreeSpace = false

      await expect(cardGenerator.generateCard(seed)).resolves.toMatchSnapshot()
    })

    it('should create a card that matches snapshot with different shuffle iteration', async () => {
      cardGenerator.options.shuffleIterations = 10

      await expect(cardGenerator.generateCard(seed)).resolves.toMatchSnapshot()
    })
  })

  describe('#classicMatch', () => {
    let cardA: Card
    let cardB: Card

    beforeEach(() => {
      // prettier-ignore
      cardA = [
        1, 2, 3, 4, 5,
        1, 2, 3, 4, 5,
        1, 2, 3, 4, 5,
        1, 2, 3, 4, 5,
        1, 2, 3, 4, 5
      ]

      cardB = [...cardA] as Card
    })

    it('should return true on identical cards', () => {
      expect(cardGenerator.classicMatch(cardA, cardB)).toBe(true)
    })

    it('should return true with a column match', () => {
      cardB.fill(15, 5, 25)

      expect(cardGenerator.classicMatch(cardA, cardB)).toBe(true)
    })

    it('should return true on row match', () => {
      // prettier-ignore
      cardB = [
        1, 12, 13, 14, 15,
        1, 12, 13, 14, 15,
        1, 12, 13, 14, 15,
        1, 12, 13, 14, 15,
        1, 12, 13, 14, 15
      ]

      expect(cardGenerator.classicMatch(cardA, cardB)).toBe(true)
    })

    it('should return true on right diagonally match', () => {
      // prettier-ignore
      cardB = [
        1, 12, 13, 14, 15,
        11, 2, 13, 14, 15,
        11, 12, 3, 14, 15,
        11, 12, 13, 4, 15,
        11, 12, 13, 14, 5
      ]

      expect(cardGenerator.classicMatch(cardA, cardB)).toBe(true)
    })

    it('should return true on left diagonally match', () => {
      // prettier-ignore
      cardB = [
        11, 12, 13, 14, 5,
        11, 12, 13, 4, 15,
        11, 12, 3, 14, 15,
        11, 2, 13, 14, 15,
        1, 12, 13, 14, 15
      ]

      expect(cardGenerator.classicMatch(cardA, cardB)).toBe(true)
    })

    it('should return false on no match', () => {
      // prettier-ignore
      cardB = [
        11, 12, 13, 14, 15,
        11, 12, 13, 14, 15,
        11, 12, 13, 14, 15,
        11, 12, 13, 14, 15,
        11, 12, 13, 14, 15
      ]

      expect(cardGenerator.classicMatch(cardA, cardB)).toBe(false)
    })
  })
})
