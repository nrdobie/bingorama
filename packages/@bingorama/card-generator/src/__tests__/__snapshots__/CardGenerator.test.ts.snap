// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardGenerator #generateCard should create a card that matches snapshot 1`] = `
Array [
  3,
  9,
  13,
  7,
  14,
  10,
  15,
  11,
  12,
  5,
  2,
  4,
  0,
  6,
  13,
  8,
  2,
  12,
  9,
  5,
  13,
  2,
  1,
  6,
  3,
]
`;

exports[`CardGenerator #generateCard should create a card that matches snapshot with different shuffle iteration 1`] = `
Array [
  5,
  8,
  15,
  10,
  11,
  1,
  13,
  11,
  3,
  12,
  11,
  10,
  0,
  5,
  13,
  8,
  7,
  11,
  4,
  1,
  13,
  1,
  10,
  5,
  15,
]
`;

exports[`CardGenerator #generateCard should create a card that matches snapshot with no free space 1`] = `
Array [
  3,
  9,
  13,
  7,
  14,
  10,
  15,
  11,
  12,
  5,
  2,
  4,
  6,
  13,
  10,
  8,
  2,
  12,
  9,
  5,
  13,
  2,
  1,
  6,
  3,
]
`;

exports[`CardGenerator #generateCards should fail if too many duplicates 1`] = `
Array [
  Array [
    1,
    2,
    3,
    4,
    5,
    1,
    2,
    3,
    4,
    5,
    1,
    2,
    3,
    4,
    5,
    1,
    2,
    3,
    4,
    5,
    1,
    2,
    3,
    4,
    5,
  ],
]
`;

exports[`CardGenerator #generateCards should generate a card set 1`] = `
Array [
  Array [
    3,
    9,
    13,
    7,
    14,
    10,
    15,
    11,
    12,
    5,
    2,
    4,
    0,
    6,
    13,
    8,
    2,
    12,
    9,
    5,
    13,
    2,
    1,
    6,
    3,
  ],
  Array [
    11,
    1,
    3,
    4,
    12,
    10,
    4,
    15,
    7,
    6,
    10,
    2,
    0,
    7,
    9,
    15,
    13,
    3,
    9,
    5,
    1,
    8,
    10,
    2,
    7,
  ],
]
`;

exports[`CardGenerator #generateCards should remove duplicates 1`] = `
Array [
  Array [
    1,
    2,
    3,
    4,
    5,
    1,
    2,
    3,
    4,
    5,
    1,
    2,
    3,
    4,
    5,
    1,
    2,
    3,
    4,
    5,
    1,
    2,
    3,
    4,
    5,
  ],
  Array [
    11,
    12,
    13,
    14,
    15,
    11,
    12,
    13,
    14,
    15,
    11,
    12,
    13,
    14,
    15,
    11,
    12,
    13,
    14,
    15,
    11,
    12,
    13,
    14,
    15,
  ],
]
`;
